### ISSUE TITLE
conda is slow

### ISSUE THREAD

**Issue by dabreegster** [Urban-Analytics/RAMP-UA#301]
Hi, I've just joined the Urban Analytics team at ATI! I'm ramping up (all puns intentional) on this codebase (the `Ecotwins-withCommuting` branch particularly) and hope to make some contributions.

To start with, I'm having an awfully hard time with the initial `conda env create -f environment.yml` setup -- its been stuck in SAT-solving the dependencies for about 3 hours now, on a reasonably powerful machine. After reading a bit about [conda performance problems](https://www.anaconda.com/blog/understanding-and-improving-condas-performance), I started shopping around for other Python dependency management solutions, and started playing around with [poetry](https://python-poetry.org/). One reason I'm attracted to it is that it uses lockfiles by default to record exactly what versions of every dependency are used. Conda, without the use of [another package](https://github.com/conda-incubator/conda-lock), doesn't. Without lockfiles, it means anybody setting up the RAMP code on a new machine could wind up using different versions of all the packages listed in `environment.yml` when they don't pin to an exact version. This would mean different people working on the same code wind up running _different_ code, which is kind of a nightmare for reproducibility and debugging.

So in short, I'm going to try to switch from conda to poetry and see what the performance and developer usability impacts are. Would that be a welcome change?

**Comment by nickmalleson** [comment]
Welcome @dabreegster ! Yes I think that would be welcome, you're not alone in having trouble setting up the environment initially with conda. I'm also keen to test whether it will work with docker, but haven't gotten round to it (my main concern is whether the opencl parts of the code will be able to access the hardware properly).

**Comment by dabreegster** [comment]
> my main concern is whether the opencl parts of the code will be able to access the hardware properly

Good news, no problems at all! I just successfully ran the OpenCL model with my Intel GPU using `poetry run python main_model.py -p model_parameters/default.yml`.

I'll continue with cutting over Ecotwins to poetry.

**Comment by cheginit** [comment]
I just stumbled upon this issue and thought I share my 2 cents! I've had great success with using `mamba`. It's a drop-in replacement for `conda` and is very fast! I use [Mambaforge](https://github.com/conda-forge/miniforge) which install packages from `conda-forge` by default so no more `-c conda-forge --strict-channel-priority`!

**Comment by dabreegster** [comment]
> I've had great success with using `mamba`

Thanks for the pointer! It certainly seems like a strictly better option than `conda`, but unless it produces lockfiles by default, there's still the issue of dependencies changing for different people based on when they first create the environment. Any reason to prefer conda/mamba over poetry, which doesn't appear to have any performance problems at all and creates lockfiles?

Another reason to be suspicious of the conda setup -- in this repo, we're importing `requests` and `sklearn`, but those dependencies aren't declared in `environment.yml`. They just happen to be transitive dependencies of some other package, so it works. This doesn't work in `poetry` -- I had to explicitly declare the dependency, which seems desirable.

**Comment by cheginit** [comment]
The main reason that I use `conda` is the need for some hard-to-install-from-`pip` packages such as `rasterio`. Many of the geospatial packages that I use depend on `gdal` which its installation is nightmare when it comes to cross-platform development. All packages on `conda-forge` are pre-compiled so installing them is straight forward from a UX perspective. I agree that `poetry` provides automation and more control over dependency management and might be a better choice for cases with dependencies that are easy to install on all platforms.

BTW, There's another package that I use for generating `meta.yml` files called [grayskull](https://github.com/conda-incubator/grayskull). You can use the `run` section of the generated `meta.yml` as a starting point for creating `environment.yml`.

**Comment by dabreegster** [comment]
Cleaning up this old issue. For future reference, we got poetry working just fine for this project: https://github.com/dabreegster/uatk-aspics/blob/main/pyproject.toml
